ports/stm32/pyb_irq.c

```diff
@@ -0,0 +1,103 @@
+/*
+ * This file is part of the MicroPython project, http://micropython.org/
+ *
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2018 Tobias Badertscher
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include <stdio.h>
+#include <string.h>
+#include <stdarg.h>
+
+
+#include "py/runtime.h"
+#include "py/gc.h"
+#include "py/mphal.h"
+#include "py/mperrno.h"
+#include "pyb_irq.h"
+
+
+void irq_dispatch(pyb_irq_obj_t *self) {
+    if (self->handler != mp_const_none) {
+        mp_sched_lock();
```

Comment:
I could imaging as well to return the flags (self-flags) in the call to the handler function. 
The the handling of the IRQ can then be decoupled from the next IRQ. Otherwise if the IRQ handling is scheduled and a new IRQ comes in the flags are cleared and the pre-IRQ scheduled function gets the wrong flags.
For consitency both soft/hard IRQ handler function should get the flags as parameter:
```
def callback(self, irq_flags, args...):
       ....
```

Reply:
Yes I did think about this as well, to pass irq_flags into the callback.  One (big?) problem with this is that you then need to create a tuple to hold all the args to the callback because `mp_sched_schedule` can only take 1 argument for the callback.  And you can't allocate a new tuple in a hard IRQ handler so it would need to be statically allocated somewhere, then you anyway run into the issue of overwriting the old data if the static allocation runs out.

There is also other state associated with an IRQ (eg trigger value) and I think it keeps it simple to just pass the IRQ object itself to the callback handler, and then the callback can query the IRQ object for all the needed state (irq_flags, trigger, etc).

I think it's fair that if the callback handler can't process IRQs fast enough then only the most recent flags would be available.  You can anyway only register one callback per IRQ object (eg uart.irq(...), pin.irq(...)) so that handler should always be able to handle all the registered trigger values.

There's always the option to use a hard IRQ handler if you need to mask new IRQs while handling the current one.